src/flags.zig

```diff
@@ -233,3 +396,409 @@ fn default_value(comptime field: std.builtin.Type.StructField) ?field.field_type
+            const argv = [_][]const u8{
+                zig_exe,
+                "build-exe",
+                "--main-pkg-path",
+                src_dir,
+                src_dir ++ "/testing/flags_test_program.zig",
+            };
+            const exec_result = try std.ChildProcess.exec(.{
+                .allocator = gpa,
+                .argv = &argv,
+                .cwd_dir = tmp_dir.dir,
+            });
+            defer gpa.free(exec_result.stdout);
+            defer gpa.free(exec_result.stderr);
+
+            if (exec_result.term.Exited != 0) {
+                std.debug.print("{s}{s}", .{ exec_result.stdout, exec_result.stderr });
+                return error.FailedToCompile;
+            }
+
+            const flags_exe = try tmp_dir.dir.realpath(
+                "flags_test_program" ++ if (builtin.os.tag == .windows) ".exe" else "",
+                flags_exe_buf,
+            );
+
+            const sanity_check = try std.fs.openFileAbsolute(flags_exe, .{});
+            sanity_check.close();
+
+            return .{
+                .gpa = gpa,
+                .tmp_dir = tmp_dir,
+                .output_buf = output_buf,
+                .flags_exe_buf = flags_exe_buf,
+                .flags_exe = flags_exe,
+            };
+        }
+
+        fn deinit(t: *T) void {
+            t.gpa.destroy(t.flags_exe_buf);
+            t.output_buf.deinit();
+            t.tmp_dir.cleanup();
+            t.* = undefined;
+        }
+
+        fn check(t: *T, cli: []const []const u8, want: Snap) !void {
+            const argv = try t.gpa.alloc([]const u8, cli.len + 1);
+            defer t.gpa.free(argv);
+
+            argv[0] = t.flags_exe;
+            for (argv[1..]) |*arg, i| {
+                arg.* = cli[i];
+            }
+            if (cli.len > 0) {
+                assert(argv[argv.len - 1].ptr == cli[cli.len - 1].ptr);
+            }
+
+            const exec_result = try std.ChildProcess.exec(.{
+                .allocator = t.gpa,
+                .argv = argv,
+            });
+            defer t.gpa.free(exec_result.stdout);
+            defer t.gpa.free(exec_result.stderr);
+
+            t.output_buf.clearRetainingCapacity();
+
+            if (exec_result.term.Exited != 0) {
+                try t.output_buf.writer().print("status: {}\n", .{exec_result.term.Exited});
+            }
+            if (exec_result.stdout.len > 0) {
+                try t.output_buf.writer().print("stdout:\n{s}", .{exec_result.stdout});
+            }
+            if (exec_result.stderr.len > 0) {
+                try t.output_buf.writer().print("stderr:\n{s}", .{exec_result.stderr});
+            }
+
+            try want.diff(t.output_buf.items);
+        }
+    };
+
+    var t = try T.init(std.testing.allocator);
+    defer t.deinit();
+
+    // Test-cases are roughly in the source order of the corresponding features.
+
+    try t.check(&.{"empty"}, snap(@src(),
+        \\stdout:
+        \\empty
+        \\
+    ));
```

Comment:
This actually is a great motivational example for snapshot testing! I can imaging us wanting to tweak various error messages over time, and now its trivial to update this. 

I also thought about using `shell.zig` to compile&run the test executable, to create a singularity out of our three micro-libraries, but decided not to. The "run" bit is not a usual shell-scripting, because there failure is very much expected, and we want to get everything there is about process exectuion, which `std.ChildProcess.exec` gives you. `shell.zig` would've work for compiling bit, but decided to be locally consistent instead.